File: merge_sort.rb
def merge_sort(list)
  if list.length < 2
    list
  else
    mid = (list.length / 2).floor
    left_array = merge_sort(list[0..mid - 1])
    right_array = merge_sort(list[mid..list.length])
    merge(left_array, right_array)
  end
end

def merge(left, right)
  if left.empty?
    right
  elsif right.empty?
    left
  elsif left[0] < right[0]
    [left.first] + merge(left[1..left.length], right)
  else
    [right.first] + merge(left, right[1..right.length])
  end
end

arr = [12,3,2,1,5,7,9]
print merge_sort(arr)

File: mergeSort.js
// The mergeSort function breaks down a single array into two smaller array elements unless it is a one element array.
// If the array is 1 element long then we return the array.
function mergeSort(arr){
  if (arr.length < 2) {
    console.log(arr)
    return arr
  }

  var middle = Math.floor(arr.length / 2)
  var leftArray = arr.slice(0, middle)
  var rightArray = arr.slice(middle, arr.length)
  console.log(leftArray)
  console.log(rightArray)
  return merge(mergeSort(leftArray), mergeSort(rightArray))
}

// The merge function takes two arrays and stitches them together forming 1 new sorted array
function merge(leftArray, rightArray){
  var newArray = []

  while (leftArray.length > 0 && rightArray.length > 0){
    if (leftArray[0] <= rightArray[0]){
      newArray.push(leftArray.shift())
    } else {
      newArray.push(rightArray.shift())
    }
  }

  while (leftArray.length > 0){
    newArray.push(leftArray.shift())
  }

  while (rightArray.length > 0){
    newArray.push(rightArray.shift())
  }

  console.log(newArray) // Watch the order in the recursion play through
  return newArray
}

console.log("Sorted Array: [" + mergeSort([2, 5, 3]) + "]")
// Uncomment addtional arrays here to play around:
// console.log("Sorted Array: [" + mergeSort([2, 4, 1, 3]) + "]")
// console.log("Sorted Array: [" + mergeSort([8, 2, 12, 4, 1, 3]) + "]")
